core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalServiceHandler.java

```diff
@@ -11,36 +11,153 @@
+        () -> mLastSnapshotUploadSize);
+  }
+
+  @Override
+  public void requestLatestSnapshotInfo(LatestSnapshotInfoPRequest request,
+                                        StreamObserver<SnapshotMetadata> responseObserver) {
+    LOG.debug("Received request for latest snapshot info");
+    if (Context.current().isCancelled()) {
+      responseObserver.onError(
+          Status.CANCELLED.withDescription("Cancelled by client").asRuntimeException());
+      return;
+    }
+    SnapshotInfo snapshot = mStateMachineStorage.getLatestSnapshot();
+    SnapshotMetadata.Builder metadata = SnapshotMetadata.newBuilder();
+    if (snapshot == null) {
+      LOG.debug("No snapshot to send");
+      metadata.setExists(false);
+    } else {
+      LOG.debug("Found snapshot {}", snapshot.getTermIndex());
+      metadata.setExists(true)
+          .setSnapshotTerm(snapshot.getTerm())
+          .setSnapshotIndex(snapshot.getIndex());
+    }
+    responseObserver.onNext(metadata.build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void requestLatestSnapshotData(SnapshotMetadata request,
+                                     StreamObserver<SnapshotData> responseObserver) {
+    if (Context.current().isCancelled()) {
+      responseObserver.onError(
+          Status.CANCELLED.withDescription("Cancelled by client").asRuntimeException());
+      return;
+    }
+    Timer.Context time = MetricsSystem.timer(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_SNAPSHOT_UPLOAD_TIMER.getName()).time();
+
+    TermIndex index = TermIndex.valueOf(request.getSnapshotTerm(), request.getSnapshotIndex());
+    String snapshotDirName = SimpleStateMachineStorage
+        .getSnapshotFileName(request.getSnapshotTerm(), request.getSnapshotIndex());
+    Path snapshotPath = new File(mStateMachineStorage.getSnapshotDir(), snapshotDirName).toPath();
+
+    try (OutputStream snapshotOutStream = new OutputStream() {
+      long mTotalBytesSent = 0L;
+      final byte[] mBuffer = new byte[mSnapshotReplicationChunkSize];
+      int mBufferPosition = 0;
+
+      @Override
+      public void write(int b) {
+        mBuffer[mBufferPosition++] = (byte) b;
+        if (mBufferPosition == mBuffer.length) {
+          flushBuffer();
+        }
+      }
+
+      @Override
+      public void close() {
+        if (mBufferPosition > 0) {
+          flushBuffer();
+        }
+        mLastSnapshotUploadDuration = time.stop() / 1_000_000L; // to get a value in ms
+        LOG.debug("Total bytes sent: {}", mTotalBytesSent);
+        MetricsSystem.histogram(
+                MetricKey.MASTER_EMBEDDED_JOURNAL_SNAPSHOT_UPLOAD_HISTOGRAM.getName())
+            .update(mTotalBytesSent);
+        mLastSnapshotUploadSize = mTotalBytesSent;
+        LOG.info("Uploaded snapshot {}", index);
+      }
+
+      private void flushBuffer() {
+        // avoids copy
+        ByteString bytes = UnsafeByteOperations.unsafeWrap(mBuffer, 0, mBufferPosition);
```

Comment:
I think you can use this unsafe here, but then you should not reuse mBuffer, and create a new one instead